File: src/main/java/net/cjsah/console/command/source/CommandSource.java
package net.cjsah.console.command.source;

import net.cjsah.console.exceptions.CommandException;
import net.cjsah.console.plugin.Plugin;
import org.apache.logging.log4j.Level;

public abstract class CommandSource<T> {
    protected final T source;

    protected CommandSource(final T source) {
        this.source = source;
    }

    public abstract boolean CanUse(Plugin plugin);

    public abstract void sendFeedBack(String message) throws CommandException;

    public abstract void sendFeedBack(String message, Level level) throws CommandException;

    public T getSource() {
        return this.source;
    }
}

File: src/main/kotlin/net/cjsah/console/command/source/CommandSource.kt
package net.cjsah.console.command.source

import net.cjsah.console.Permissions.PermissionType
import net.cjsah.console.exceptions.CommandException
import net.cjsah.console.plugin.Plugin
import org.apache.logging.log4j.Level

abstract class CommandSource<T>(private val source: T) {

    abstract fun hasPermission(permission: PermissionType): Boolean

    abstract fun canUse(plugin: Plugin): Boolean

    @Throws(CommandException::class)
    abstract fun sendFeedBack(message: String)

    @Throws(CommandException::class)
    abstract fun sendFeedBack(message: String, level: Level)

    open fun getSource(): T {
        return source
    }

}